File: pbark/pbark/parser/controlflow/function_parser.py
from __future__ import annotations

from pbark.errors import BarkError
from pbark.lexer import TokenType
from pbark.parser import ast_node as AstNode
from pbark.parser import keywords as Keywords
from pbark.parser.expression.value_parser import ValueParser


class FunctionParser:
    def __init__(self, parser) -> None:
        self._parser = parser
        self._value_parser = ValueParser(parser)

    def parse_trick_line(self):
        if not self._parser.line_has_word(Keywords.EXPECTS_WORDS):
            return None
        line = self._parser.peek().line
        expects_at = self._find_word(Keywords.EXPECTS_WORDS)
        if expects_at < 0:
            return None
        name = self._find_name_before(expects_at)
        if name is None:
            raise BarkError(line, "Every trick needs a name")
        returns_at = self._find_word_after(Keywords.TRICK_RETURNS_WORDS, expects_at + 1)
        if returns_at < 0:
            raise BarkError(line, f'Trick "{name}" never says what the dogs get out of it!')
        params = self._read_params(expects_at + 1, returns_at)
        stop_at = self._find_word_after(Keywords.BURY_KEYWORDS, returns_at + 1)
        value_end = stop_at if stop_at >= 0 else self._line_end_offset()
        return_expr = self._value_parser.parse_part(returns_at + 1, value_end)
        if stop_at >= 0:
            self._parser.advance_by(stop_at + 1)
            self._parser.skip_newlines()
            steps = []
        else:
            self._parser.consume_until_line_end()
            self._parser.skip_newlines()
            steps = self._read_body_until_stop()
            if not self._parser.current_line_is_bury():
                raise BarkError(line, "The trick started but you forgot to say stop!")
            self._parser.consume_bury_line()
        return AstNode.FunctionDef(name, params, steps, return_expr, line)

    def _find_name_before(self, expects_at: int) -> str | None:
        for offset in range(expects_at):
            token = self._parser.peek_at(offset)
            if token.is_(TokenType.IDENTIFIER):
                return self._parser.normalise(token.value)
        return None

    def _read_params(self, from_: int, to: int) -> list[str]:
        params = []
        for offset in range(from_, to):
            token = self._parser.peek_at(offset)
            if not token.is_(TokenType.IDENTIFIER):
                continue
            word = self._parser.normalise(token.value)
            if word in Keywords.EXPECTS_WORDS or word in Keywords.TRICK_RETURNS_WORDS:
                continue
            if word in Keywords.LIST_AND_WORDS:
                continue
            params.append(word)
        return params

    def _line_end_offset(self) -> int:
        offset = 0
        while not self._parser.is_at_end_or_line_end_offset(offset):
            offset += 1
        return offset

    def _read_body_until_stop(self):
        steps = []
        while not self._parser.is_at_end():
            self._parser.skip_newlines()
            if self._parser.is_at_end() or self._parser.current_line_is_bury():
                break
            self._parser.parse_statement_into(steps)
        return steps

    def _find_word(self, words: list[str]) -> int:
        return self._find_word_after(words, 0)

    def _find_word_after(self, words: list[str], from_: int) -> int:
        offset = from_
        while not self._parser.is_at_end_or_line_end_offset(offset):
            token = self._parser.peek_at(offset)
            if token.is_(TokenType.IDENTIFIER) and self._parser.normalise(token.value) in words:
                return offset
            offset += 1
        return -1

File: src/main/java/dev/klomptech/jbark/parser/controlflow/FunctionParser.java
package dev.klomptech.jbark.parser.controlflow;

import dev.klomptech.jbark.errors.BarkError;
import dev.klomptech.jbark.lexer.Token;
import dev.klomptech.jbark.lexer.TokenType;
import dev.klomptech.jbark.parser.AstNode;
import dev.klomptech.jbark.parser.Keywords;
import dev.klomptech.jbark.parser.ParseExpression;
import dev.klomptech.jbark.parser.Parser;
import dev.klomptech.jbark.parser.expression.ValueParser;
import java.util.ArrayList;
import java.util.List;

public class FunctionParser {

  private final Parser parser;
  private final ValueParser valueParser;

  public FunctionParser(final Parser parser) {
    this.parser = parser;
    this.valueParser = new ValueParser(parser);
  }

  // Example: add expects a b returns a plus b bury
  public AstNode parseTrickLine() throws BarkError {
    if (!parser.lineHasWord(Keywords.EXPECTS_WORDS)) {
      return null;
    }
    int line = parser.peek().line();
    int expectsAt = findWord(Keywords.EXPECTS_WORDS);
    if (expectsAt < 0) {
      return null;
    }
    String name = findNameBefore(expectsAt);
    if (name == null) {
      throw new BarkError(line, "Every trick needs a name");
    }
    int returnsAt = findWordAfter(Keywords.TRICK_RETURNS_WORDS, expectsAt + 1);
    if (returnsAt < 0) {
      throw new BarkError(line, "Trick \"" + name + "\" never says what the dogs get out of it!");
    }
    List<String> params = readParams(expectsAt + 1, returnsAt);
    int stopAt = findWordAfter(Keywords.BURY_KEYWORDS, returnsAt + 1);
    int valueEnd = stopAt >= 0 ? stopAt : lineEndOffset();
    ParseExpression returnExpr = valueParser.parsePart(returnsAt + 1, valueEnd);
    List<AstNode> steps;
    if (stopAt >= 0) {
      parser.advanceBy(stopAt + 1);
      parser.skipNewlines();
      steps = List.of();
    } else {
      parser.consumeUntilLineEnd();
      parser.skipNewlines();
      steps = readBodyUntilStop();
      if (!parser.currentLineIsBury()) {
        throw new BarkError(line, "The trick started but you forgot to say stop!");
      }
      parser.consumeBuryLine();
    }
    return new AstNode.FunctionDef(name, params, steps, returnExpr, line);
  }

  private String findNameBefore(final int expectsAt) {
    for (int offset = 0; offset < expectsAt; offset++) {
      Token token = parser.peekAt(offset);
      if (token.is(TokenType.IDENTIFIER)) {
        return parser.normalise(token.value());
      }
    }
    return null;
  }

  private List<String> readParams(final int from, final int to) {
    List<String> params = new ArrayList<>();
    for (int offset = from; offset < to; offset++) {
      Token token = parser.peekAt(offset);
      if (!token.is(TokenType.IDENTIFIER)) {
        continue;
      }
      String word = parser.normalise(token.value());
      if (Keywords.EXPECTS_WORDS.contains(word) || Keywords.TRICK_RETURNS_WORDS.contains(word)) {
        continue;
      }
      if (Keywords.LIST_AND_WORDS.contains(word)) {
        continue;
      }
      params.add(word);
    }
    return params;
  }

  private int lineEndOffset() {
    int offset = 0;
    while (!parser.isAtEndOrLineEndOffset(offset)) {
      offset++;
    }
    return offset;
  }

  private List<AstNode> readBodyUntilStop() throws BarkError {
    List<AstNode> steps = new ArrayList<>();
    while (!parser.isAtEnd()) {
      parser.skipNewlines();
      if (parser.isAtEnd() || parser.currentLineIsBury()) {
        break;
      }
      parser.parseStatementInto(steps);
    }
    return steps;
  }

  private int findWord(final List<String> words) {
    return findWordAfter(words, 0);
  }

  private int findWordAfter(final List<String> words, final int from) {
    for (int offset = from; !parser.isAtEndOrLineEndOffset(offset); offset++) {
      Token token = parser.peekAt(offset);
      if (token.is(TokenType.IDENTIFIER) && words.contains(parser.normalise(token.value()))) {
        return offset;
      }
    }
    return -1;
  }
}
